python/pyspark/sql/pandas/conversion.py

```diff
@@ -297,8 +297,11 @@ class SparkConversionMixin(object):
     """
     Min-in for the conversion from pandas to Spark. Currently, only :class:`SparkSession`
     can use this class.
+    pandasRDD=True creates a DataFrame from an RDD of pandas dataframes
+    (currently only supported using arrow)
```

Comment:
Would it make sense to do type checking here instead of having a flag?

Reply:
Can we somehow define/get the type of the RDD[py-object] without evaluating the first element of it?
If not, then the RDD might contain any type of object, so the _pandasRDD_ option is used as a way to differentiate between initialization from an RDD and an RDD of pd.DataFrames.

Thank you for reviewing! please let me know if there's anything else i can do to get this merged.

Reply:
That's a good point. If we look in `session.py` we can see _createFromRDD does it magic there. Personally I would put this logic instead inside of `_inferSchema` and `toInternal` respectively but I'm coming at this from more a core-spark dev perspective maybe @HyukjinKwon has a different view.

Reply:
Even if we don't refactor this back into session.py, I'd encourage you to look at session.py and consider structuring this in a similar way so that we don't have to have this flag here.

Reply:
I agree that this seems to fit well into `_inferSchema` & `_createFromRDD`, although we still would need some way to discern between an rdd of DataFrames and other types when the user provides a schema (and we don't want to peek into the first item).

Do you think it would be better to move the pandas flag into `_createFromRDD`?

Reply:
So _inferSchema does effectively peek into the first element. I think we could just put the logic down inside of the map and then the user doesn't have to specify this flag.

Reply:
That's in case the user wants to infer the schema (so we have to peek into the rdd), but in case the use does specify the schema, there's no need to peek, and we're left with no other option to tell which code path we need

Reply:
So let's say the user specifies a schema, in that case inside of _createFromRDD we can just look at the type of each element that were processing and see if it's a DataFrame or a Row or a Dictionary and dispatch the logic there. What do you think? Or is there a reason I'm missing why we couldn't do the dispatch inside of _createFromRDD based on type?

Reply:
Well in case the user specifies a schema, the entire process is lazy, so there's no need to evaluate any of the rdd elements...

if we keep everything lazy and map each element to either a row or RecordBatch, we would still need to know which path to take, e.g. for RecordBatches we need to call:
```python
        from pyspark.sql.dataframe import DataFrame
        jrdd = rb_rdd._to_java_object_rdd()
        jdf = self._jvm.PythonSQLUtils.toDataFrame(jrdd, schema.json(), self._wrapped._jsqlContext)
        df = DataFrame(jdf, self._wrapped)
        df._schema = schema
        return df
```
and for Rows we need to call:
```python
        jrdd = self._jvm.SerDeUtil.toJavaArray(rdd._to_java_object_rdd())
        jdf = self._jsparkSession.applySchemaToPythonRDD(jrdd.rdd(), schema.json())
        df = DataFrame(jdf, self._wrapped)
        df._schema = schema
        return df
```